fem/integ/bilininteg_hcurl_kernels.hpp

```diff
@@ -0,0 +1,3869 @@
+               }
+               for (int dy = 0; dy < D1Dy; ++dy)
+               {
+                  const double wy = Bct(dy,qy);
+                  const double wDy = Gct(dy,qy);
+
+                  for (int dz = 0; dz < D1Dz; ++dz)
+                  {
+                     gradYZ01[dz][dy] += wy * massZ[dz][1];
+                     gradYZ10[dz][dy] += wDy * massZ[dz][0];
+                  }
+               }
+            }
+
+            for (int dx = 0; dx < D1Dx; ++dx)
+            {
+               const double wx = Bct(dx,qx);
+               const double wDx = Gct(dx,qx);
+
+               for (int dy = 0; dy < D1Dy; ++dy)
+               {
+                  for (int dz = 0; dz < D1Dz; ++dz)
+                  {
+                     // \hat{\nabla}\times\hat{u} is [(u_2)_{x_1}, -(u_2)_{x_0}, 0]
+                     // (u_2)_{x_1} * (op * curl)_0 - (u_2)_{x_0} * (op * curl)_1
+                     Y(dx + ((dy + (dz * D1Dy)) * D1Dx) + osc,
+                       e) += (gradYZ10[dz][dy] * wx) - (gradYZ01[dz][dy] * wDx);
+                  }
+               }
+            }
+         }  // loop qx
+      }
+   }); // end of element loop
+}
+
+template<int MAX_D1D = HCURL_MAX_D1D, int MAX_Q1D = HCURL_MAX_Q1D>
+inline void SmemPACurlCurlApply3D(const int D1D,
+                                  const int Q1D,
+                                  const bool symmetric,
+                                  const int NE,
+                                  const Array<double> &bo,
+                                  const Array<double> &bc,
+                                  const Array<double> &bot,
+                                  const Array<double> &bct,
+                                  const Array<double> &gc,
+                                  const Array<double> &gct,
+                                  const Vector &pa_data,
+                                  const Vector &x,
+                                  Vector &y)
+{
+   MFEM_VERIFY(D1D <= MAX_D1D, "Error: D1D > MAX_D1D");
+   MFEM_VERIFY(Q1D <= MAX_Q1D, "Error: Q1D > MAX_Q1D");
+   // Using (\nabla\times u) F = 1/det(dF) dF \hat{\nabla}\times\hat{u} (p. 78 of Monk), we get
+   // (\nabla\times u) \cdot (\nabla\times v) = 1/det(dF)^2 \hat{\nabla}\times\hat{u}^T dF^T dF \hat{\nabla}\times\hat{v}
+   // If c = 0, \hat{\nabla}\times\hat{u} reduces to [0, (u_0)_{x_2}, -(u_0)_{x_1}]
+   // If c = 1, \hat{\nabla}\times\hat{u} reduces to [-(u_1)_{x_2}, 0, (u_1)_{x_0}]
+   // If c = 2, \hat{\nabla}\times\hat{u} reduces to [(u_2)_{x_1}, -(u_2)_{x_0}, 0]
+
+   auto Bo = Reshape(bo.Read(), Q1D, D1D-1);
+   auto Bc = Reshape(bc.Read(), Q1D, D1D);
+   auto Gc = Reshape(gc.Read(), Q1D, D1D);
+   auto op = Reshape(pa_data.Read(), Q1D, Q1D, Q1D, symmetric ? 6 : 9, NE);
+   auto X = Reshape(x.Read(), 3*(D1D-1)*D1D*D1D, NE);
+   auto Y = Reshape(y.ReadWrite(), 3*(D1D-1)*D1D*D1D, NE);
+
+   const int s = symmetric ? 6 : 9;
+
+   auto device_kernel = [=] MFEM_HOST_DEVICE (int e)
```

Comment:
```suggestion
   auto device_kernel = [=] MFEM_DEVICE (int e)
```

This (and the same change made in other places) is the cause of the warnings referenced above.

@sebastiangrimberg, do you know of a good way to do a full diff of the moved function bodies to make sure that inadvertent edits like this don't get merged?

Reply:
Looks like I was too aggressive with my find-and-replace in [9c0eae5](https://github.com/mfem/mfem/pull/3526/commits/9c0eae5e3e43f3f80d34ac8cbdfa9ab56b6215f2). Fixed this in [1a614d5](https://github.com/mfem/mfem/pull/3526/commits/1a614d51a2d98b9c0d13458e2339fcac9f41ffa3).

It would be nice to compare function-by-function changes in this branch, but unfortunately no I don't know any way to do it easily. I just parse the full diff to check my work and up to this point have done this pretty stringently to make sure things are correct. The mistake from [1a614d5](https://github.com/mfem/mfem/pull/3526/commits/1a614d51a2d98b9c0d13458e2339fcac9f41ffa3) is really touching anything inside of functions but I understand the desire to double check everything. I will give things another review now just to make extra sure, but I guess at the end of the day we will have to trust strong unit testing + CI testing.

Reply:
Thanks for the fix. It builds cleanly with CUDA for me now… I'll re-merge in next.

Please do check over the changes again and see if anything else slipped through the cracks.